man/flute.man.1
.\" Copyright (c) 1999-2003 INRIA - Universite Paris 6 - All rights reserved
.\" (main author: Vincent Roca - vincent.roca@inrialpes.fr)
.\" 
.\"  This program is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\" 
.\"  This program is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"
.\"  You should have received a copy of the GNU General Public License
.\"  along with this program; if not, write to the Free Software
.\"  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
.\"  USA.


.TH FLUTE "1" "" "MCLv3 Reference Manual"

.SH NAME

flute \- send or receive files, as specified in FLUTE (RFC 3926) and using
the ALC transport protocol (RFC 3452).

.SH SYNOPSIS

.B flute
.nh
[\fB-h\fP[\fBelp\fP]]
[\fB-send\fP]
[\fB-recv\fP]
[\fB-a address\fP[\fB/port\fP]]
[\fB-ifa address\fP]
[\fB-ifn ifname\fP]
[\fB-demux demux_level\fP]
[\fB-v verbosity_level\fP]
[\fB-stat stat_level\fP]
[\fB-silent\fP]
[\fB-tmp tmp_dir_string\fP]
.\" [\fB-objaggr\fP]
[\fB-o speed/space/cpu\fP]
[\fB-p low|med|high|lan\fP]
[\fB-p size\fP[\fB/rate\fP]]
[\fB-l nb_of_layers\fP]
[\fB-cc congestion_control_protocol\fP]
[\fB-P\fP]
[\fB-t ttl\fP]
[\fB-R\fP]
[\fB-cont\fP]
[\fB-repeat n\fP]
[\fB-fec n\fP]
.\" [\fB-huge (DEPRECATED)\fP]
[\fB-src address\fP
[\fB-int\fP]
[\fBfile or directory\fP]
.hy

.PP
At a sender:
.nf
.B flute -send [options] file
.fi
.nf
.B flute -send [options] -R directory
.fi

.PP
At a receiver:
.nf
.B flute -recv [options]
.fi


.SH DESCRIPTION

.B flute
is a multicast file distribution application built over the 
MultiCast Library (MCLv3) which uses the ALC (Asynchronous
Layered Coding) reliable multicast protocol.
.B flute
can run both as a sender or a receiver.
.B flute
has many features like the ability to send recursively all the files
of a given directory and its subdirectories, and the possibility to be used
in 
.B ``on-demand''
or
.B ``push''
mode.
.B flute
communicates many meta-data information along with the files in
associated File Delivery Tables (FDT), in order to let the receiver
choose what files he wants to receive.
An application-level MD5 checksum is also used in order to detect
possible transmission errors.

.B fcast
and
.B flute
differ from the fact that the first one uses private, non standard,
solutions, while the second one strictly follows IETF standards.
.B fcast 
is appropriate when looking for a simple yet efficient tool, while
.B flute
is appropriate when using IETF standardized solutions is mandatory, or
when a large set of files are sent within a session and receivers
may be interested only in a subset of them.
.B flute
filtering mechanism is in that case a real asset.
The two tools are otherwise sharing many similarities, as well
as most arguments.


.SH OPTIONS

The following arguments can be provided
.B both at a sending or receiving flute.

.TP
.B \-h[elp]
Gives the credits, the compilation parameters and all the possible 
.B flute
arguments.

.TP
.B \-send or \-recv
Set the Fcast application mode to 
.B sender 
OR to 
.B receiver. 
You must ALWAYS specify one of these two modes.

.TP
.B \-a address[/port]
Specifies the unicast or multicast IPv4 or IPv6 address of the base layer and its UDP
port number (if specified).
If no
.B -a
and/or
.B /port
argument is provided, then the default target address (usually that of
the loopback interface) and port number of MCL are used.
With multicast addresses, the address and port number of layers 1 and above
are derived from that of the base layer by MCL (usually by incrementing the
address and port number).
.I (communicated to MCL)

.TP
.B \-ifa address
The network interface to use is the one attached to the local address
specified (or hostname specified). 
This option is only required on multi-homed hosts or on routers as the
default interface chosen by the system without this option to send/receive
multicast may not be the one you want to use.
The alternative is to use
.B \-ifn ifname.
.I (communicated to MCL)

.TP
.B \-ifn ifname
The network interface to use is the one attached to the name
specified (e.g. 
.B \-ifneth0
).
This option is only required on multi-homed hosts or on routers as the
default interface chosen by the system without this option to send/receive
multicast may not be the one you want to use.
The alternative is to use
.B \-ifa address.
.I (communicated to MCL)

.TP
.B \-demux N
Sets the LCT Transport Session Identifier (or TSI) to 
.B N. 
At a sender this TSI is included in each packet sent.
At a receiver <src_addr; TSI> are used to fully identify the ALC session
and to filter incoming packets.
If no
.B -demux
argument is provided, then any packet matching the source address (if
provided) will be accepted with the risk they belong to other sessions
(e.g if they use the same <multicast address, port> tuple).
This demux label is also usefull to make sure that the current session will
not get confused by packets from previous incarnation session.
The receivers and the sender must of course agree beforehand and use the
same value.
By default a value of 0 is used.
.I (communicated to MCL)

.B \-v verbosity_level
Specifies the level of verbosity desired from the MCL library.
This option also automatically enables the \-stat 2 option to
add statistics.
Note that full verbosity requires that the MCL library be compiled in
DEBUG mode.
.I (communicated to MCL)

.TP
.B \-stat stat_level
Specifies the level of statistics desired from the MCL library.
Level 0 (default) means nothing, level 1 means that only the final
statistics (at end of session) will be displayed, level 2 means that each
time an object is fully received, statistics are displayed.
.I (communicated to MCL)

.TP
.B \-silent
Enables the silent mode (nothing is sent to stdout).

.TP
.B \-tmp tmp_dir_string
Set the MCL temporary directory to be tmp_dir_string rather than its
default (usually "/tmp" with Unix).
The argument is a string, e.g. :
-tmp/home/roca/tmp
.I (communicated to MCL)

.TP
.B \-objaggr
Enables the object aggregation mode.
This mode is highly efficient when sending a large number of small objects
(i.e. files) since FEC encoding now operates on the large object (that is the
result of the aggregation of all the small objects), rather than on
each small object.
Transmissions are an order of magnitude more efficient.
This solution is similar in spirit to the one described in
<draft-neumann-rmt-flute-file-aggregation-00.txt>.
This mode requires that the recursive mode is set (to transmit several files).
.I (communicated to MCL)

.TP
.B \-o space/speed/cpu
Specifies the optimization profile (in particular the order in which packets
are sent on each layer).
Possible values are:
.B space 
to optimize the maximum memory space required at a receiver,
.B speed
to optimize the reception speed, and
.B cpu
to differ CPU intensive tasks at a receiver (recommanded on a 
CPU bounded host).
If not specified, the default value of MCL is used instead.
.I (communicated to MCL)

.I WARNING: it is important that the transmission and reception optimization match, so use it on both ends.


.PP

The following arguments define the transmission profiles.
One can use either a predefined profile, or do everything by hand,
or redefine some parameters of the predefined profiles.
These arguments must be specified at the sender and/or at the receiver,
as explained.

.TP
.B \-p low|med|high|lan
Specifies the transmission profile. 
Possible values are:
.B low 
(low rate Internet), 
.B med 
(mid rate Internet),
.B high
(high speed Internet), and
.B lan
(high speed LAN).
If not specified, the default value of MCL is used instead.
The
.B -p size[/rate]
argument can be used instead (but not at the same time).
The
.B -plan
profile is dedicated to LAN environments, which means that no
congestion control is used (CCI field of the LCT header left to zero, which
also means that no loss statistics are possible),
and a single ALC layer is defined. Use with care...

The exact parameters associated to each of these four profiles
can be changed either (1) with the
.B -p size[/rate]
or
.B -l number_of_layers
or
.B -cc congestion_control_protocol
arguments,
or (2) by modifying the src/alc/mcl_tx_prof.cpp file.
.I (communicated to MCL)

.I WARNING: it is important that the transmission and reception profiles match, so specify the profile on both ends.

.TP
.B \-p size[/rate]
Specifies the datagram size (in bytes) (at a sender or receiver), and
optionally the transmission rate (in bits/s) (only at a sender).
This argument cannot be used along with one of the
.B \-p low|med|high|lan
arguments.
.I (communicated to MCL)

.I WARNING: it is important that the transmission and reception profiles match, so specify it on both ends (only -p size at a receiver).


.TP
.B \-l number_of_layers
Specifies the maximum number of layers to use.
If not specified, the default value of MCL is used instead.
Note that a receiver is anyway constrained by the number of layers
offered by the source.
.I (communicated to MCL)

.TP
.B \-cc congestion_control_protocol
Specifies the congestion control protocol to use.
.B -cc0
specifies that NO CONGESTION CONTROL is used, and a single
layer is selected.
Use this mode only in appropriate environments, not over the Internet
where congestion control is mandatory.
This argument replaces the now deprecated "singlelayer" mode.
.B -cc1
specifies RLC congestion control (often needed for interoperability tests).
.B -cc2
specifies FLID-SL congestion control. This is the default, since this
protocol is less aggressive than RLC.
If nothing is specified, the default value of MCL is used instead.
.I (communicated to MCL)

.TP
.B \-singlelayer (DEPRECATED)
This argument is DEPRECATED and replaced by the -cc0 argument,
which must be used when no congestion control is required.
.TP

.TP
.B \-P
If enabled, causes flute to make a pause before exiting (WIN32 only).
Used to prevent the closing of console window when flute is done, thus
giving some time to read outputs.


.PP

The following arguments are
.B specific to a sender.


.TP
.B \-t ttl
Specifies the time to live (ttl) value of the IP header.
If not specified, the default value of MCL is used instead (usually 1).
Use \-t 1 if all the receivers are attached to the same LAN as the source.
.I (communicated to MCL)

.TP
.B \-R
Enables the recursive mode.
Using this flag, flute recursively sends the whole directory tree given as a
parameter.
At the receiver side, the same directory tree is created in the CURRENT 
directory (flute never goes upward).

.TP
.B \-cont
Set the
.I continuous mode
for "on-demand" transmissions (i.e. receivers arrive
at their discretion, download the file and leave).
By default it is not set.
.I (communicated to MCL)

.TP
.B \-repeat n

doc/RELEASE_NOTES
	$Id: RELEASE_NOTES,v 1.29 2005/05/24 15:43:21 roca Exp $

	-----------------------------------------------------
	MCL LIBRARY RELEASE NOTES - vincent.roca@inrialpes.fr
	-----------------------------------------------------


Supported functionalities
-------------------------

The following table lists the supported functionalities in the
various releases...

  Functionalities                     library version number
                                        2.3  2.4 2.99 2.991 2.992 2.993 2.994 2.995 2.996
------upper interface-----------------|----|----|----|-----|-----|-----|-----|-----|-----|
scatter/gather support                | NO | NO | NO | NO  | NO  | NO  | P   | P   | Y   |
FLUTE API                             | NO | NO | NO | NO  | NO  | NO  | NO  | NO  | Y   |
                                      |    |    |    |     |     |     |     |     |     |
------networking----------------------|----|----|----|-----|-----|-----|-----|-----|-----|
IPv6 support                          | NO | NO | NO | NO  | NO  | NO  | P   | Y   | Y   |
IPv4 SSM (Source Specific Multicast)  | NO | NO | NO | NO  | NO  | NO  | NO  | NO  | Y   |
IPv6 SSM (Source Specific Multicast)  | NO | NO | NO | NO  | NO  | NO  | NO  | NO  | P   |
                                      |    |    |    |     |     |     |     |     |     |
------reliability---------------------|----|----|----|-----|-----|-----|-----|-----|-----|
FEC (reed-solomon) for reliability    | Y  | Y  | Y  | Y   | Y   | Y   | Y   | Y   | Y   |
LDGM large block FEC codec            | NO | NO | Y  | Y   | Y   | Y   | Y   | Y   | Y   |
                                      |    |    |    |     |     |     |     |     |     |
------ALC/LCT functionalities---------|----|----|----|-----|-----|-----|-----|-----|-----|
ALC (Asynch Layered Coding) compliant | Y  | Y  | Y  | Y   | R   | Y   | Y   | Y   | Y   |
LCT (Layered Coding Transp) compliant | Y  | Y  | Y  | Y   | E   | Y   | Y   | Y   | Y   |
virtual transmission memory service   | Y  | Y  | Y  | NO  | M   | NO  | NO  | NO  | NO  |
virtual reception memory service      | Y  | Y  | Y  | NO  | O   | NO  | NO  | NO  | Y   |
congestion control (RLC)              | Y  | Y  | Y  | Y   | V   | Y   | Y   | Y   | Y   |
congestion control (FLID-SL)          | Y  | Y  | Y  | Y   | E   | Y   | Y   | Y   | Y   |
select() support                      | P  | P  | Y  | Y   | D   | Y   | Y   | Y   | Y   |
1->n with feedback n->1               | Y  | Y  | NO | NO  | !   | NO  | NO  | NO  | NO  |
m->n support (can be both tx and rx)  | NO | NO | NO | NO  | !   | NO  | NO  | NO  | NO  |
support of file size >> memory size   | Y  | Y  | NO | NO  | !   | NO  | NO  | NO  | NO  |
                                      |    |    |    |     |     |     |     |     |     |
------NORM functionalities------------|----|----|----|-----|-----|-----|-----|-----|-----|
NORM support included                 | NO | NO | NO | NO  | Y   | Y   | Y   | Y   | Y   |
NORM compliant                        | NO | NO | NO | NO  | NO  | NO  | NO  | NO  | NO  |
congestion control (pgmcc/tfmcc)      | NO | NO | NO | NO  | NO  | NO  | NO  | NO  | NO  |
                                      |    |    |    |     |     |     |     |     |     |
------applications--------------------|----|----|----|-----|-----|-----|-----|-----|-----|
flute compliant                       | NO | NO | NO | P   | NO  | Y   | Y   | Y   | Y   |
fcast application                     | Y  | Y  | Y  | Y   | Y   | Y   | Y   | Y   | Y   |
windows fcast GUI frontend            | Y  | Y  | Y  | Y   | Y   | Y   | Y   | NO  | NO  |
robcast application                   | NO | NO | Y  | Y   | Y   | Y   | Y   | Y   | Y   |
                                      |    |    |    |     |     |     |     |     |     |
------system--------------------------|----|----|----|-----|-----|-----|-----|-----|-----|
linux support                         | Y  | Y  | Y  | Y   | Y   | Y   | Y   | Y   | Y   |
solaris support                       | Y  | Y  | Y  | Y   | Y   | Y   | Y   | Y   | Y   |
freebsd support                       | NO | NO | NO | P   | NO  | P   | Y   | Y   | Y   |
windows XP (visual C++)               | Y  | Y  | Y  | Y   | NO  | P   | P   | Y   | Y   |
                                      |    |    |    |     |     |     |     |     |     |

Y: yes		(validated)
P: partial	(not fully validated or including known bugs or important
		 limitations)
NO: not available


Notes for MCLv3 release 2.99.6 - May, 2005.
-----------------------------------------------------

This is a major version with many changes for the ALC and
FLUTE part:
- Creation of a C++ FLUTE API. The Flute application is now a small application
  built on top of the FLUTE API. Writting your own application on top of this
  FLUTE API enables a full and dynamic control of the FLUTE/ALC session, e.g. to
  add transmission cycles to the carousel, add/remove files to the carousel, and
  many other things. Extending the example Flute application to add new
  functionalities is both feasible and straightforward.
- Receivers are no longer limited by their RAM size (receiving files whose
  size is an order of magnitude larger than the host's RAM is feasible).
  This is made possible by a dedicated virtual receive memory management scheme.
- IPv4 SSM (source specific multicast) is now supported.
- IPv6 SSM (source specific multicast) is now supported (Linux with 2.6.x
  kernels, tested with FC2 and FC3 distributions).
- Added automatic MCLv3-FLUTE versus MAD-FLUTE interoperability validation
  scripts.


2/ NORM part:
- No change.

Dependancies: requires LDPC v1.8; only tested with Xerces-C++ v2.6.0



src/flute_lib/Flute.cpp
/* $Id: Flute.cpp,v 1.7 2005/05/24 09:56:15 moi Exp $ */
/*
 *  Copyright (c) 2003-2004 INRIA - All rights reserved
 *  main authors: Christoph Neumann - christoph.neumann@inrialpes.fr
 *		  Vincent Roca - vincent.roca@inrialpes.fr
 *		  Julien Laboure - julien.laboure@inrialpes.fr
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
 *  USA.
 */

#include "flute_includes.h"


/**
 * Flute constructor
 */
Flute::Flute () {

	flutecb = new flute_cb();

}


/**
 * Flute destructor
 */
Flute::~Flute () {

	/* KILL threads */

	if (flutecb->id != 0) mcl_close(flutecb->id);
	delete flutecb;

}


/**
 * Abort the current Flute session
 */
void Flute::abort () {

	if (flutecb->id != 0) mcl_abort(flutecb->id);
	
}


/**
 * Print an mcl about (showing compile options, version, etc.)
 */
void Flute::about () {

	flutecb->lock();

	if (mcl_ctl(flutecb->id, MCL_OPT_MOREABOUT, NULL, 0))
		EXIT(("Flute: Usage: mcl_ctl failed for MOREABOUT\n"))
		
	flutecb->unlock();
}


/**
 * Sets the mcl verbosity level
 * @param verbosity 	verbosity level
 */
void Flute::setVerbosity (INT32 verbosity) {

	flutecb->lock();
	
	flutecb->verbose = verbosity;
	flutecb->stats = 2;

	flutecb->unlock();

}


/**
 * Get the mcl verbosity level used
 * @return 	verbosity level
 */
INT32 Flute::getVerbosity () {

	INT32 result;
	flutecb->lock();
	result = flutecb->verbose;
	flutecb->unlock();
	return result;

	
}


/**
 * Sets the mcl stat level
 * @param stats 	statistic level
 */
void Flute::setStatsLevel (INT32 stats) {

	flutecb->lock();

	flutecb->stats = stats;

	flutecb->unlock();

}


/**
 * Get the mcl stat level used
 * @return 	statistic level
 */
INT32 Flute::getStatsLevel () {

	INT32 result;
	flutecb->lock();
	result = flutecb->stats;
	flutecb->unlock();
	return result;
	
}


/**
 * Sets the congestion control scheme:
 *	INVALID_CC:	initial value, nothing choosen
 *	NO_CC:		choose not to have any CC
 *	RLC_CC:		choose to have RLC CC
 *	FLID_SL_CC:	choose to have FLID SL (static layer) CC
 * @param cc_scheme	the congestion control scheme
 */
void Flute::setCCScheme (INT32 cc_scheme) {

	flutecb->lock();

	flutecb->congestion_control = cc_scheme;
	if (flutecb->congestion_control == NO_CC && flutecb->nb_layers == 0) flutecb->nb_layers = 1;

	flutecb->unlock();

}


/**
 * Sets the transmission profile:
 *	MCL_TX_PROFILE_LOW_RATE_INTERNET:	modem connection
 *	MCL_TX_PROFILE_MID_RATE_INTERNET:	eg with VPNs, xDSL
 *	MCL_TX_PROFILE_HIGH_SPEED_INTERNET:	several Mbps connection
 *	MCL_TX_PROFILE_HIGH_SPEED_LAN
 * @param tx_profile	the transmission profile
 */
void Flute::setTxProfile (INT32 tx_profile) {

	flutecb->lock();

	flutecb->txprof_mode = tx_profile;
	if (flutecb->txprof_mode == MCL_TX_PROFILE_HIGH_SPEED_LAN)
		flutecb->congestion_control = NO_CC; /* true in a LAN! */
	flutecb->txprof_set = true;

	flutecb->unlock();
	
}


/**
 * Sets the datagram size in bytes.
 * @param dt_size	datagram size in bytes
 */
void Flute::setDatagramSize (INT32 dt_size) {

	flutecb->lock();

	flutecb->txprof_dt_size = dt_size;

	flutecb->unlock();

}


/**
 * Sets the tx rate (in bits/sec).
 * @param tx_rate	 transmission rate in bits/s
 */
void Flute::setTxRate (INT32 tx_rate) {

	flutecb->lock();

	if (tx_rate > 0) {
		flutecb->txprof_rate = (double)tx_rate /
			((double)flutecb->txprof_dt_size * 8.0);
		/* no less than 1 pkt/s */
		flutecb->txprof_rate = max(1.0, flutecb->txprof_rate);
	}

	flutecb->unlock();

}


/**
 * Sets the nb of layers used.
 * @param nb_layers	 the number of layers
 */
void Flute::setNbLayers (INT32 nb_layers) {

	flutecb->lock();

	flutecb->nb_layers = nb_layers;

	flutecb->unlock();

}


/**
 * Sets the IP version used.
 * @param ip_version	 the ip version
 */
#ifdef WIN32
void Flute::setIpVersion (INT32 ip_version) {

	flutecb->lock();

	flutecb->ip_version = ip_version;
	if (flutecb->ip_version != 4 && flutecb->ip_version !=6)
	{
		EXIT(("Flute: ERROR, unknown IP version %i. Specify IP version 4 or 6\n",ip_version))
	}

	flutecb->unlock();
}
#endif


/**
 * Sets the SSM (source specific multicast) mode on/off.
 * @param set	 on or off?
 */
void Flute::setSSM (bool set) {

	flutecb->lock();

	flutecb->ssm = set;

	flutecb->unlock();
}


/**
 * Sets the session address on which to send/recv.
 * @param session_addr	 session address
 */
void Flute::setSessionAddr (struct sockaddr *session_addr) {

	flutecb->lock();
	
	flutecb->session_addr = session_addr;
	
	flutecb->unlock();
}


/**
 * Sets the session address on which to send/recv.
 * @param session_addr_str	 session address string
 */
void Flute::setSessionAddr (char	*session_addr_str) {

	flutecb->lock();

	strncpy(flutecb->session_addr_str, session_addr_str, sizeof(flutecb->session_addr_str));
	flutecb->session_address_set = true;

	flutecb->unlock();

}


/**
 * Sets the port of the Flute session
 * @param port	port number of the session
 */
void Flute::setPort (INT32 port) {

	flutecb->lock();

	flutecb->port = port;

	flutecb->unlock();

}


#ifndef WIN32
/**
 * Sets the interface to be used for rx and tx.
 * @param if_addr_str	 interface address string
 */
void Flute::setInterface (char *if_addr_str) {

	flutecb->lock();

	if (flutecb->tools->extract_addr(if_addr_str, &flutecb->mcast_if_addr, 0) < 0) {
		EXIT(("Flute: setInterface: ERROR, address extraction failed\n"))
	}

	flutecb->unlock();

}


/**
 * Sets the interface to be used for rx and tx.
 * @param if_name	 interface address
 */
void Flute::setInterface (struct sockaddr *if_addr) {

	flutecb->lock();
	
	flutecb->mcast_if_addr = if_addr;
	
	flutecb->unlock();

}


/**
 * Sets the interface to be used for rx and tx.
 * @param if_name	 interface name
 */
void Flute::setInterfaceName (char* if_name) {

	flutecb->lock();

	flutecb->mcast_if_name = (char *) calloc(1, strlen(if_name) + 1);
	if (flutecb->mcast_if_name == NULL) {
		EXIT(("Flute: ERROR, calloc failed during Flute::setInterfaceName\n"))
	}
	strncpy(flutecb->mcast_if_name, if_name, strlen(if_name) + 1);

	flutecb->unlock();

}
#endif /* !WIN32 */


/**
 * Sets the TSI of the Flute session
 * @param tsi	tsi of the session
 */
void Flute::setTSI (INT32 tsi) {

	flutecb->lock();

	flutecb->demux_label = tsi;

	flutecb->unlock();

}


/**
 * Set the temp dir (where temporary files are stored).
 * @param tmp_dir	name of the temp dir
 */
void Flute::setTmpDir (char * tmp_dir) {

	flutecb->lock();

	strncpy(flutecb->tmp_dir, tmp_dir, sizeof(flutecb->tmp_dir));
	flutecb->tmp_dir[MAX_PATH + MAX_FILENAME -1] = 0;
	/* add final / and \0 if not already present */
	if (tmp_dir[strlen(flutecb->tmp_dir)-1] != '/' &&
		    strlen(tmp_dir) <= MAX_PATH + MAX_FILENAME -2)
		    strcat(flutecb->tmp_dir, "/");
 	flutecb->tmp_dir_set = true;

	flutecb->unlock();

}


/**
 * Creartes an sdp file using the parameters of the current Flute session.
 * @param	filename of the sdp file
 */
void Flute::createSdp (char* fileparam) {

	flutecb->lock();

	flutecb->sdp = new FluteSDP(this->flutecb);
	flutecb->sdp->generateSDP(fileparam);
	delete flutecb->sdp;
	flutecb->sdp = NULL;

	flutecb->unlock();

}


/**
 * Parses an sdp file and sets all parameters accordingly.
 * @param	filename of the sdp file
 */
void Flute::parseSdp (char* fileparam) {

	flutecb->lock();

	flutecb->sdp = new FluteSDP(this->flutecb);
	flutecb->sdp->parseSDP(fileparam);
	delete flutecb->sdp;
	flutecb->sdp = NULL;

	flutecb->unlock();
}


/**
 * Get the port of the Flute session.
 * @return 	port of the flute session
 */
INT32 Flute::getPort () {

	INT32 val;
	
	flutecb->lock();
	val = flutecb->port;
	flutecb->unlock();

	return val;

}


/**
 * Are we are flute sender?
 * @return	true or false
 */
bool Flute::isSender() {


	flutecb->lock();
	
	if (flutecb->mode == SEND) {
		
		flutecb->unlock();
		return true;
	}
	
	flutecb->unlock();
	return false;
	
}


/**
 * Are we are flute receiver?
 * @return	true or false
 */
bool Flute::isReceiver() {

	flutecb->lock();
	
	if (flutecb->mode == RECV) {
		flutecb->unlock();
		return true;
	}
	flutecb->unlock();
	return false;
}


/**
 * Returns file information of the given toi
 * @param toi	toi of the file
 * @return	a FluteFileInfo class containing info about the file
 *              NULL if no file with given toi exists
 */
class FluteFileInfo *Flute::getFileInfo (TOI_t toi) 
{

	class FluteFileInfo* returnedFFileInfo;
	flutecb->lock();
	returnedFFileInfo = flutecb->fdt->getFileInfo(toi);
	flutecb->unlock();
	
	return returnedFFileInfo;
	
}


/**
 * Returns a list of file information. It is a list of files
 * contained in the FDT.
 * @return	a list of FluteFileInfo classes 
 *              containing info about the files
 *		NULL if FDT is empty.
 */
class FluteFileInfo *Flute::getFileInfoList () 
{
	class FluteFileInfo* returnedFFileInfo;
	flutecb->lock();
	returnedFFileInfo = flutecb->fdt->getFileInfoList();
	flutecb->unlock();	
	
	return returnedFFileInfo;

}


/**
 * Initialize the ALC library now.
 * All selected options are passed to MCL.
 */
void Flute::init ()
{
	int err = 0;

	ASSERT(flutecb->mode == SEND || flutecb->mode == RECV);

	/* Process address string */
	if (flutecb->session_address_set == true &&
	    flutecb->tools->extract_addr(flutecb->session_addr_str, &flutecb->session_addr, flutecb->port) < 0) {
			EXIT(("Flute: ERROR, extract_addr failed for session addr argument -a%s\n", flutecb->session_addr_str))
	}

	/* Some sanity checks */
	if (flutecb->congestion_control == NO_CC && flutecb->nb_layers > 1)
		EXIT(("Flute: ERROR, cannot set nb of layers > 1 with no congestion control\n"))
#ifndef WIN32		
	if (flutecb->session_addr != NULL && flutecb->session_addr->sa_family == AF_INET6 && flutecb->mcast_if_name == NULL)
	{
		EXIT(("Flute: ERROR, you must specify an interface name with IPv6\n"))
	}
#endif
	if (flutecb->ssm && (flutecb->src_addr == NULL) && (flutecb->mode == RECV)) {
		EXIT(("Flute: ERROR, you must specify a source address in SSM/Receiver mode\n"))
	}

	/* specify few important parameters... */
	err += mcl_ctl(flutecb->id, MCL_OPT_VERBOSITY, (void*)&flutecb->verbose, sizeof(flutecb->verbose));
	EXIT_ON_ERROR(err, ("Flute: ERROR, mcl_ctl failed for VERBOSITY\n"))
	err += mcl_ctl(flutecb->id, MCL_OPT_STATS,
				(void*)&flutecb->stats, sizeof(int));
	EXIT_ON_ERROR(err, ("Flute: ERROR, mcl_ctl failed for STATS\n"))
	err += mcl_ctl(flutecb->id, MCL_OPT_SET_FLUTE_MODE, NULL, 0);
	EXIT_ON_ERROR(err, ("Flute: ERROR, mcl_ctl failed for SET_FLUTE_MODE\n"))
	if (flutecb->congestion_control > 0) {	/* must be before TX_PROFILE */
		err += mcl_ctl(flutecb->id, MCL_OPT_SET_CC_SCHEME,
				(void*)&flutecb->congestion_control,
				sizeof(flutecb->congestion_control));
		EXIT_ON_ERROR(err, (
			"Flute: ERROR, mcl_ctl failed for MCL_OPT_SET_CC_SCHEME %d\n",
			flutecb->congestion_control))
	}
	if (flutecb->txprof_set) {
		err += mcl_ctl(flutecb->id, MCL_OPT_TX_PROFILE, (void*)&flutecb->txprof_mode,
				sizeof(flutecb->txprof_mode));
		EXIT_ON_ERROR(err, (
			"Flute: ERROR, mcl_ctl failed for TX_PROFILE\n"))
	}
	if (flutecb->txprof_dt_size > 0) {	/* must be after TX_PROFILE */
		err += mcl_ctl(flutecb->id, MCL_OPT_DATAGRAM_SIZE,
				(void*)&flutecb->txprof_dt_size,
				sizeof(flutecb->txprof_dt_size));
		EXIT_ON_ERROR(err, (
			"Flute: ERROR, mcl_ctl failed for DATAGRAM_SIZE %d\n",
			flutecb->txprof_dt_size))
	}
	if (flutecb->txprof_rate > 0.0) {	/* must be after TX_PROFILE */
		err += mcl_ctl(flutecb->id, MCL_OPT_TX_RATE,
				(void*)&flutecb->txprof_rate,
				sizeof(flutecb->txprof_rate));
		EXIT_ON_ERROR(err, (
			"Flute: ERROR, mcl_ctl failed for TX_RATE %f\n",
			flutecb->txprof_rate))
	}
	if (flutecb->nb_layers > 0) {
		err += mcl_ctl(flutecb->id, MCL_OPT_LAYER_NB,
				(void*)&flutecb->nb_layers,
				sizeof(int));
		EXIT_ON_ERROR(err, ("Flute: ERROR, mcl_ctl failed for LAYER_NB\n"))
	}
	if (flutecb->session_addr != NULL) {
		int	addr_len;
		addr_len = (flutecb->session_addr->sa_family == AF_INET6) ?
				sizeof(struct sockaddr_in6) :
				sizeof(struct sockaddr_in);
		err += mcl_ctl(flutecb->id, MCL_OPT_BIND,
				(void*)flutecb->session_addr, addr_len);
		EXIT_ON_ERROR(err, ("Flute: ERROR, mcl_ctl failed for BIND\n"))
	}
	if (flutecb->mcast_if_addr != NULL) {		/* in host format! */
		err += mcl_ctl(flutecb->id, MCL_OPT_SET_NETIF_ADDR,
				(void*)&flutecb->mcast_if_addr,
				sizeof(*flutecb->mcast_if_addr));
		EXIT_ON_ERROR(err, ("Flute: ERROR, mcl_ctl failed for SET_NETIF_ADDR\n"))
	} else if (flutecb->mcast_if_name != NULL) {
		err += mcl_ctl(flutecb->id, MCL_OPT_SET_NETIF_NAME,
				(void*)flutecb->mcast_if_name,
				strlen(flutecb->mcast_if_name));
		EXIT_ON_ERROR(err, ("Flute: ERROR, mcl_ctl failed for SET_NETIF_NAME\n"))
	}
	if (flutecb->ssm == true) {
		int	val = (int)flutecb->ssm;
		err += mcl_ctl(flutecb->id, MCL_OPT_SET_SSM,
				(void*)&val, sizeof(val));
		EXIT_ON_ERROR(err, (
			"Flute: ERROR, mcl_ctl failed for DEMUX_LABEL\n"))
	}
	if (flutecb->demux_label > 0) {
		err += mcl_ctl(flutecb->id, MCL_OPT_DEMUX_LABEL,
				(void*)&flutecb->demux_label,
				sizeof(flutecb->demux_label));
		EXIT_ON_ERROR(err, (
			"Flute: ERROR, mcl_ctl failed for DEMUX_LABEL\n"))
	}
	if (flutecb->ttl >= 0) {
		err += mcl_ctl(flutecb->id, MCL_OPT_TTL,
				(void*)&flutecb->ttl,
				sizeof(flutecb->ttl));
		EXIT_ON_ERROR(err, (
			"Flute: ERROR, mcl_ctl failed for TTL\n"))
	}
	if (flutecb->tmp_dir_set) {
		err += mcl_ctl(flutecb->id, MCL_OPT_TMP_DIR,
				(void*)&flutecb->tmp_dir,
				strlen(flutecb->tmp_dir));
		EXIT_ON_ERROR(err, (
			"Flute: ERROR, mcl_ctl failed for TMP_DIR \"%s\"\n",
			flutecb->tmp_dir))
	}

	if (flutecb->mode == SEND) {
		/*
		 * always use LCT1 now, in all cases...
		 */
		int mcl_option = MCL_SCHED_LCT1;
		if (mcl_ctl(flutecb->id, MCL_OPT_SCHED,
					(void*)&mcl_option,
					sizeof(mcl_option))) {
			EXIT(("Flute: mcl_ctl MCL_OPT_SCHED failed for LCT1\n"))
		}
	
	}
	else if (flutecb->mode == RECV) {
		if (flutecb->src_addr != NULL) {
			int	addr_len;
			addr_len = (flutecb->src_addr->sa_family == AF_INET6) ?
				sizeof(struct sockaddr_in6) :
				sizeof(struct sockaddr_in);
			if (mcl_ctl(flutecb->id, MCL_OPT_SRC_ADDR,
					(void*)flutecb->src_addr, addr_len)) {
				EXIT(("mcl_ctl: MCL_OPT_SRC_ADDR failed\n"))
			}
		}
	
	}
	flutecb->initialized = true;
}

